packages/flutter_tools/lib/src/protocol_discovery.dart

```diff
@@ -9,71 +9,83 @@ import 'base/port_scanner.dart';
 import 'device.dart';
 import 'globals.dart';
 
-/// Discovers a specific service protocol on a device, and forward the service
+/// Discovers a specific service protocol on a device, and forwards the service
 /// protocol device port to the host.
 class ProtocolDiscovery {
   ProtocolDiscovery._(
-    DeviceLogReader logReader,
-    String serviceName, {
+    this.logReader,
+    this.serviceName, {
     this.portForwarder,
     this.hostPort,
     this.defaultHostPort,
-  }) : _logReader = logReader, _serviceName = serviceName {
-    assert(_logReader != null);
+  }) : _prefix = '$serviceName listening on ' {
+    assert(logReader != null);
     assert(portForwarder == null || defaultHostPort != null);
-    _deviceLogSubscription = _logReader.logLines.listen(_onLine);
+    _deviceLogSubscription = logReader.logLines.listen(_handleLine);
+    _timer = new Timer(const Duration(seconds: 60), () {
+      _stopScrapingLogs();
+      _completer.completeError(new ToolExit('Timeout while attempting to retrieve URL for $serviceName'));
+    });
+  }
+
+  factory ProtocolDiscovery.observatory(
+    DeviceLogReader logReader, {
+    DevicePortForwarder portForwarder,
+    int hostPort,
+  }) {
+    const String _kObservatoryService = 'Observatory';
```

Comment:
No need for the underscore anymore (or below in the `diagnosticService` factory constructor)